distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso

```diff
@@ -1,59 +1,476 @@
+    group_nums_to_names : Map Integer Text
+    group_nums_to_names self =
+        map = polyglot_map_to_map self.internal_regex_object.groups
+        map.transform k-> v-> [v, k]
+
+    ## Compile the provided `expression` into a regex pattern that can be used for
+       matching.
+
+       Arguments
+       - expression: The text representing the regular expression that you want to
+         compile. Must be non-empty.
+       - case_insensitive: Enables or disables case-insensitive matching. Case
+         insensitive matching behaves as if it normalises the case of all input
+         text before matching on it.
+
+       If an empty regex is used, `compile` throws an Illegal_Argument error.
+
+       ? Why Compile?
+         While many regex engines are able to cache ad-hoc patterns, it is often
+         useful to be able to manually retain a pattern that you have computed. This
+         function exists so you can hold onto the resultant `Regex` object,
+         instead of immediately proceeding to match using it.
+    compile : Text -> Boolean | Nothing -> Regex ! Regex_Syntax_Error | Illegal_Argument
+    compile expression case_insensitive=Nothing =
+        if expression == '' then Error.throw (Illegal_Argument.Error "Regex cannot be the empty string") else
+            options_string = if case_insensitive == True then "usgi" else "usg"
+
+            internal_regex_object = Panic.catch Syntax_Error (Prim_Text_Helper.compile_regex expression options_string) caught_panic->
+                Error.throw (Regex_Syntax_Error.Error (caught_panic.payload.message))
+
+            Regex.Value internal_regex_object
+
+    ## Escape the special characters in `expression` such that the result is a
+       valid literal pattern for the original string.
+
+       Arguments:
+       - expression: The expression to escape metacharacters in.
+
+       > Example
+         Turn a Text into a regex that matches that string exactly.
+
+             example_escape =
+                 literal_string = "\!\.|abcde"
+                 Regex.escape literal_string
+    escape : Text -> Text
+    escape expression = Regex_Utils.regexQuote expression
+
+## PRIVATE
+
+    Performs the regex match, and iterates through the results. Yields both
+    the matched parts of the string, and the 'filler' parts between them.
+
+    The 'filler' elements are `Utf_16_Span`s, not `Spans`. This is because
+    matches and replacement boundaries can fall in the middle of multi-
+    character graphemes, thereby splitting them apart.
+
+    At each step, it yields a Match_Iterator_Value, whivch has either a filler
+    and a match, or just the final filler. A Match_Iterator_Value.Last value is
+    return at the end, and only at the end.
+
+    Optionally, you can call `early_exit` to have it return the remainder of
+    the string, unmatched, as a single Last value. (Used for `replace` with
+    `only_first=True`.)
+type Match_Iterator
```

Comment:
I'm wondering if we should move types like this to `Internal`.

At some point definitely will. May be worth to do now, but not necessary - up to you.

Reply:
Done